Mathlib/Computability/Language.lean

```diff
@@ -0,0 +1,311 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fox Thomson
+
+! This file was ported from Lean 3 source module computability.language
+! leanprover-community/mathlib commit a239cd3e7ac2c7cde36c913808f9d40c411344f6
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.Hom.Ring
+import Mathlib.Algebra.Order.Kleene
+import Mathlib.Data.List.Join
+import Mathlib.Data.Set.Lattice
+
+/-!
+# Languages
+
+This file contains the definition and operations on formal languages over an alphabet. Note strings
+are implemented as lists over the alphabet.
+The operations in this file define a [Kleene algebra](https://en.wikipedia.org/wiki/Kleene_algebra)
+over the languages.
+-/
+
+
+open List Set Computability
+
+universe v
+
+variable {α β γ : Type _}
+
+/-- A language is a set of strings over an alphabet. -/
+def Language (α) :=
+  Set (List α)
+#align language Language
+
+instance : Membership (List α) (Language α) := ⟨Set.Mem⟩
+instance : Singleton (List α) (Language α) := ⟨Set.singleton⟩
+instance : Insert (List α) (Language α) := ⟨Set.insert⟩
+instance : CompleteBooleanAlgebra (Language α) := Set.instCompleteBooleanAlgebraSet
+
+namespace Language
+
+variable {l m : Language α} {a b x : List α}
+
+-- Porting note: `reducible` attribute cannot be local.
+attribute [reducible] Language
```

Comment:
This is a big problem; it means we now have a global instance on `Zero (Language α)`

Reply:
I created the PR #2307 to remove this.